File: modules/kasane-document/src/package.cpp
// SPDX-License-Identifier: MIT
#include <kasane/package.hpp>
#include <png.h>
#include <kasane/filesystem.hpp>
#include <random>
#include <sstream>

namespace kasane {
namespace {
std::string json_string(const std::string &value) {
    std::string s = "\"";
    const char *hex = "0123456789abcdef";
    for (unsigned char c : value) {
        if (c == '"' || c == '\\') {
            s += '\\';
            s += char(c);
        } else if (c < 32) {
            s += "\\u00";
            s += hex[c >> 4];
            s += hex[c & 15];
        } else
            s += char(c);
    }
    return s + '"';
}

} // namespace

Status publish_package(const Document &doc, const PackageOptions &options) {
    namespace fs = std::filesystem;
    if (!options.validate)
        return Status::error("MISSING_VALIDATOR",
                             "Package publication requires an explicit runtime validation gate");
    Moc3Artifact artifact;
    if (auto s = encode_moc3(doc, artifact); !s.ok())
        return s;
    auto filesystem = options.filesystem ? options.filesystem : io::native_filesystem();
    fs::path stage, backup, destination;
    std::unique_ptr<io::Lock> lock;
    bool moved = false, published = false;
    try {
        destination = io::local_path(options.destination);
        if (options.destination.empty() || destination == destination.root_path() ||
            destination.filename().empty())
            return Status::error("INVALID_DESTINATION", options.destination.string());
        if (filesystem->info(destination).exists &&
            (!filesystem->info(destination).directory || filesystem->info(destination).symlink))
            return Status::error("INVALID_DESTINATION", destination.string());
        std::vector<std::vector<uint8_t>> textures;
        for (const auto &slot : artifact.textures) {
            auto source = fs::path(slot.source);
            if (source.is_relative())
                source = options.asset_root / source;
            io::Bytes bytes;
            try {
                bytes = filesystem->read(io::local_path(source));
            } catch (const io::Error &e) {
                return Status::error("MISSING_TEXTURE", slot.asset_id + ": " + e.what());
            }
            png_image image{};
            image.version = PNG_IMAGE_VERSION;
            if (!png_image_begin_read_from_memory(&image, bytes.data(), bytes.size())) {
                png_image_free(&image);
                return Status::error("INVALID_PNG", slot.asset_id);
            }
            if (image.width != slot.width || image.height != slot.height) {
                png_image_free(&image);
                return Status::error("RESOURCE_MISMATCH",
                                     slot.asset_id + ": PNG dimensions differ from metadata");
            }
            image.format = PNG_FORMAT_RGBA;
            // Resource boundary: cap decoded data, never allocate from unchecked input.
            if (uint64_t(image.width) * image.height > 268435456) {
                png_image_free(&image);
                return Status::error("CAPACITY", slot.asset_id + ": PNG exceeds 1 GiB decoded");
            }
            std::vector<uint8_t> pixels(PNG_IMAGE_SIZE(image));
            bool ok = png_image_finish_read(&image, nullptr, pixels.data(), 0, nullptr);
            png_image_free(&image);
            if (!ok)
                return Status::error("INVALID_PNG", slot.asset_id + ": incomplete PNG data");
            textures.push_back(std::move(bytes));
        }
        if (auto s = options.validate(artifact); !s.ok())
            return s;
        filesystem->create_directories(destination.parent_path());
        lock = filesystem->lock(destination.parent_path());
        stage = destination.parent_path() / (".kasane-package-" + io::unique_name());
        filesystem->create_directory_new(stage);
        filesystem->create_directories(stage / "textures");
        io::write_new(*filesystem, stage / "model.moc3", artifact.bytes);
        auto write_text = [&](const fs::path &path, const std::string &text) {
            io::write_new(*filesystem, path,
                          std::span(reinterpret_cast<const uint8_t *>(text.data()), text.size()));
        };
        write_text(stage / "model.model3.json", artifact.model3_json);
        std::ostringstream report;
        report << "{\"status\":\"passed\",\"moc_version\":5,\"source_revision\":" << doc.revision()
               << ",\"textures\":[";
        for (size_t i = 0; i < textures.size(); ++i) {
            auto &slot = artifact.textures[i];
            io::write_new(*filesystem, stage / slot.package_path, textures[i]);
            report << (i ? "," : "") << "{\"asset_id\":" << json_string(slot.asset_id)
                   << ",\"source\":" << json_string(slot.source)
                   << ",\"path\":" << json_string(slot.package_path) << ",\"width\":" << slot.width
                   << ",\"height\":" << slot.height << "}";
        }
        report << "]}\n";
        auto text = report.str();
        write_text(stage / "export-report.json", text);
        filesystem->sync_directory(stage / "textures");
        filesystem->sync_directory(stage);
        backup = stage;
        backup += "-previous";
        if (filesystem->info(destination).exists) {
            filesystem->move_new(destination, backup);
            moved = true;
        }
        filesystem->move_new(stage, destination);
        published = true;
        try {
            filesystem->sync_directory(destination.parent_path());
        } catch (const io::Error &e) {
            if (options.publication_warnings)
                options.publication_warnings->push_back(
                    std::string("Published, but directory synchronization failed: ") + e.what());
        }
        if (moved) {
            try {
                filesystem->remove(backup);
            } catch (const io::Error &) {
            }
        }
        return {};
    } catch (const std::exception &e) {
        if (moved && !published) {
            try {
                filesystem->move_new(backup, destination);
            } catch (const io::Error &) {
                return Status::error("ROLLBACK_FAILED", std::string(e.what()) +
                                                            "; previous package retained at " +
                                                            backup.string());
            }
        }
        if (!stage.empty() && !published)
            try {
                filesystem->remove(stage);
            } catch (const io::Error &) {
            }
        return Status::error("PACKAGE_IO", e.what());
    }
}
} // namespace kasane

File: crates/kasane-project/src/package.rs
use std::fs;
use std::path::PathBuf;

use kasane_core::types::Status;
use kasane_core::Document;
use kasane_moc3::{encode_moc3, Moc3Artifact};

use crate::filesystem::{self as io, FileSystem, NativeFileSystem, Publication};
use crate::store::{asset_path, read_project_asset};

pub type ArtifactValidator = Box<dyn Fn(&Moc3Artifact) -> Result<(), Status>>;

pub struct PackageOptions {
    pub asset_root: PathBuf,
    pub destination: PathBuf,
    pub validate: Option<ArtifactValidator>,
}

/// Publish only verified source bytes; callers must inspect publication warnings.
pub fn publish_package(doc: &Document, options: &PackageOptions) -> Result<Publication, Status> {
    publish_with_filesystem(doc, options, &NativeFileSystem)
}

pub(crate) fn publish_with_filesystem(
    doc: &Document,
    options: &PackageOptions,
    files: &dyn FileSystem,
) -> Result<Publication, Status> {
    if doc.transaction_active() {
        return Err(Status::error(
            "TRANSACTION_ACTIVE",
            "Commit or cancel edits first",
        ));
    }
    let validator = options.validate.as_ref().ok_or_else(|| {
        Status::error(
            "MISSING_VALIDATOR",
            "Package publication requires an explicit validation gate",
        )
    })?;
    io::reject_symlink(&options.destination)?;
    let destination = io::local_path(&options.destination)?;
    let parent = destination
        .parent()
        .filter(|_| destination.file_name().is_some())
        .ok_or_else(|| Status::error("INVALID_DESTINATION", "Cannot replace a filesystem root"))?;
    if !options.asset_root.as_os_str().is_empty()
        && io::local_path(&options.asset_root)?.starts_with(&destination)
    {
        return Err(Status::error(
            "INVALID_DESTINATION",
            "Export would replace the source project",
        ));
    }
    for id in doc.asset_order() {
        if asset_path(&options.asset_root, doc.get_asset(id).unwrap())?.starts_with(&destination) {
            return Err(Status::error(
                "INVALID_DESTINATION",
                "Export would replace a source asset",
            ));
        }
    }

    let artifact = encode_moc3(doc)?;
    // Verify even unused project assets, as in the existing DocumentSession contract.
    // The bytes checked here are the exact bytes written below; never reopen after validation.
    let mut verified = std::collections::HashMap::new();
    for id in doc.asset_order() {
        let data = read_project_asset(&options.asset_root, doc.get_asset(id).unwrap())?;
        verified.insert(id.as_str(), data.bytes);
    }
    validator(&artifact)?;
    fs::create_dir_all(parent).map_err(io::io_error)?;
    let _lock = io::lock(parent)?;
    io::reject_symlink(&destination)?;
    if destination.exists() && !destination.is_dir() {
        return Err(Status::error(
            "INVALID_DESTINATION",
            "Destination is not a directory",
        ));
    }
    let stage = io::Stage::new(parent)?;
    fs::create_dir(stage.0.join("textures")).map_err(io::io_error)?;
    files
        .write_new(&stage.0.join("model.moc3"), &artifact.bytes)
        .map_err(io::io_error)?;
    files
        .write_new(
            &stage.0.join("model.model3.json"),
            artifact.model3_json.as_bytes(),
        )
        .map_err(io::io_error)?;
    for slot in &artifact.textures {
        files
            .write_new(
                &stage.0.join(&slot.package_path),
                &verified[slot.asset_id.as_str()],
            )
            .map_err(io::io_error)?;
    }
    let report = serde_json::json!({
        "status": "passed",
        "moc_version": 5,
        "source_revision": doc.revision(),
        "textures": artifact.textures.iter().map(|slot| serde_json::json!({
            "asset_id": slot.asset_id, "source": slot.source,
            "path": slot.package_path, "width": slot.width, "height": slot.height,
        })).collect::<Vec<_>>(),
    });
    files
        .write_new(
            &stage.0.join("export-report.json"),
            report.to_string().as_bytes(),
        )
        .map_err(io::io_error)?;
    files
        .sync_directory(&stage.0.join("textures"))
        .map_err(io::io_error)?;
    files.sync_directory(&stage.0).map_err(io::io_error)?;

    // Reserve a separate directory so a failed rollback leaves a named recovery copy.
    let backup = io::Stage::new(parent)?;
    let previous = backup.0.join("previous");
    let replacing = destination.exists();
    if replacing {
        files
            .rename(&destination, &previous)
            .map_err(io::io_error)?;
    }
    if let Err(error) = files.rename(&stage.0, &destination) {
        if replacing {
            if let Err(rollback) = files.rename(&previous, &destination) {
                let recovery = backup.preserve().join("previous");
                return Err(Status::error(
                    "ROLLBACK_FAILED",
                    format!(
                        "Publication failed: {error}; rollback failed: {rollback}; recover from {}",
                        recovery.display()
                    ),
                ));
            }
        }
        return Err(io::io_error(error));
    }
    let mut publication = Publication::finish(files, parent);
    if let Err(e) = fs::remove_dir_all(&backup.0) {
        let recovery = backup.preserve();
        publication.warnings.push(format!(
            "Published; old package retained at {}: {e}",
            recovery.display()
        ));
    }
    Ok(publication)
}
